### ISSUE TITLE
Trust based on EKPub

### ISSUE THREAD

**Issue by caleblloyd** [bloomberg/spire-tpm-plugin#4]
**Describe the solution you'd like**

Currently, EKCert chain must be validated and a node is allowed.  This requires the server to have CA roots for each TPM manufacturer, and trusts all devices with a matching CA root.

I think it would also be beneficial to allow for a list of EKPub hashes to be considered.  This would mean trust could be locked down to only certain devices that are on the EKPub whitelist

One solution would be an optional server configuration, such as:

```
ek_pub_hash_path  = "/opt/spire/.data/ek_pub_hashes"
```

Each file in that directory would be an empty file with a filename being a SHA-256 hash of an authorized EK Pub.  If `ek_pub_hash_path` is supplied, a check for file exists on `{ek_pub_hash_path}/{ek_pub_hash}` would be performed and if found, considered valid.

**Describe alternatives you've considered**

Considered a datastore API for checking EK Pubs that reside in a remote datastore, but that would introduce additional complexity into this project, and could be handled out-of-band.

**Additional context**

The solution would be similar to the "Trust based on EKPub" described here: https://docs.microsoft.com/en-us/windows-server/identity/ad-ds/manage/component-updates/tpm-key-attestation#BKMK_DeploymentOverview

**Comment by pawalt** [comment]
I think this is a good idea so long as these pubhashes are consistent with what our [pubhash function](https://github.com/bloomberg/spire-tpm-plugin/blob/6012644b5bdc8f107c39d709e4538470276d0328/pkg/common/tpm_attestor.go#L128) generates. It does a SHA256 sum encoded with base64.

This isn't something I have the time to implement right now, but if you're interested in working on it, I'd be more than happy to point you in the right direction. The logic should sit at the same place as our [validEK check](https://github.com/bloomberg/spire-tpm-plugin/blob/6012644b5bdc8f107c39d709e4538470276d0328/cmd/server/tpm_attestor/tpm_attestor.go#L112).

**Comment by caleblloyd** [comment]
I have a WIP

Now I am just trying to figure out if there is a good way to write tests for this...  Need a TPM simulator or something, digging around seeing what other projects are doing

**Comment by pawalt** [comment]
I test my stuff out on a hardware TPM, but there are some instructions here on how to simulate a TPM. If they fill the tpm 2 spec properly, they should work out.
https://github.com/tpm2-software/tpm2-tools/wiki/Getting-Started#tpm-dependency

Alternatively, you could run in GCP with a shielded VM. I'd like to get CI tests on this repo, so if either of these work out well for you, I'd love to hear how they work.
https://cloud.google.com/blog/products/gcp/security-in-plaintext-use-shielded-vms-to-harden-your-gcp-workloads

**Comment by caleblloyd** [comment]
Currently looking at using the `github.com/google/go-tpm-tools/simulator` package.

Opened an issue upstream to get some cleaner integration with passing the simulator to the `attest` package: https://github.com/google/go-attestation/issues/156

**Comment by caleblloyd** [comment]
Need to figure out how to load an EK Cert into the simulator.  This project does it with bash, just need to figure out what those commands would be in Go: https://github.com/mrcdb/tpm2_ek_cert_generator

**Comment by josephlr** [comment]
We just opened https://github.com/google/go-tpm-tools/issues/85 to track adding Cert helpers to `go-tpm-tools`. Ideally, you'd just be able to do:

```go
tpm, _ := tpm2.OpenTPM("/dev/tpmrm0")
// Get the EK
key, _ := tpm2tools.EndorsementKeyECC(tpm)
// Get an crypto.Signer interface to the TPM's key
signer, _ := key.GetSigner()
// Create a Request template
template := tpm2tools.EKCertTemplate()
// Fill out specific fields
template.Name.CommonName = "foobar"
// Create the CSR
csr, _ := x509.CreateCertificateRequest(rand.Reader, template, signer)
// Get a certificate from a CA
cert := FunctionThatCallsYourCA(csr)
// Write the cert into NVData
tpm2tools.StoreEKCert(cert)
```
All of those functions are currently implemented except for `tpm2tools.StoreEKCert` and `tpm2tools.EKCertTemplate`.

And ya, GCP recently switched to giving all machines vTPM/UEFI by default, so normal CI tools that use GCE (CircleCI, CirrusCI) should just work. Note that you can't use containers, you have to use instances (as accessing the TPM requires root), so that might make the CI more expensive/slower.

**Comment by caleblloyd** [comment]
I have a follow-up PR read for this once testing is merged
